test_shim.py

```diff
@@ -0,0 +1,225 @@
+"""
+Test that the second provided package is a near-perfect shim of the first package.
+
+This means that:
+- All modules in the shim package are in the main package.
+- All modules in the main package are in the shim package.
+- Every attribute is exported at minimum, where it is defined in the main package.
+- No additional files exist in the shim package.
+
+"""
+import argparse
+import glob
+import importlib
+import importlib.util
+import logging
+import os
+import pkgutil
+import sys
+from types import ModuleType
+from typing import Dict, List, Set, Tuple, TypedDict
+
+logger = logging.getLogger(__name__)
+
+
+class AttributeErrors(TypedDict):
+    missing_attributes: List[str]
+    mismatching_attributes: List[str]
+
+
+class Response(TypedDict):
+    extra_files: List[str]
+    missing_files: List[str]
+    attributes: Dict[str, AttributeErrors]
+
+
+def test_file(base: ModuleType, shim: ModuleType) -> AttributeErrors:
+    """Test the two files against each other. Any attributes that aren't exposed by the second file are returned as an error."""
+    missing_attributes = []
+    mismatching_attributes = []
+    for attr in dir(base):
+        if attr.startswith("_"):
+            continue
+
+        base_attr = getattr(base, attr)
+        if not hasattr(base_attr, "__module__"):
+            continue
+
+        if not base_attr.__module__.startswith(base.__name__):
+            continue
+
+        if not hasattr(shim, attr):
+            logger.error(f"{shim.__name__}.{attr} is missing.")
+            missing_attributes.append(attr)
+            continue
+
+        shim_attr = getattr(shim, attr)
+        if not hasattr(shim_attr, "__module__"):
+            continue
+        if not shim_attr.__module__.startswith((base.__name__, shim.__name__)):
+            continue
+
+        # handle main.
+        # main is special, since it needs to have equal attributes, but not the same attributes
+        # we can get a near perfect result for this by comparing that the signatures, docstrings, and so forth are equal.
+        # TODO:
```

Comment:
Empty todo, or misplaced todo?

Reply:
Todo is for the above. Additionally, will be killing this tester too, as it doesn't properly test the current iteration of the shim.